File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
 
class Solution {
  int diameterOfBinaryTree(TreeNode? root) {
    int largest_diameter = 0;
    int height(TreeNode? root) {
      if (root == null) return 0;

      int right_height = height(root.right);
      int left_height = height(root.left);

      int diameter = right_height + left_height;

      largest_diameter = max(largest_diameter, diameter);

      return 1 + max(right_height, left_height);
    }

    height(root);

    return largest_diameter;
  }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_road = 0
        
        def get_hight(node: Optional[TreeNode]) -> int:
            nonlocal max_road  # <-- This is the key fix
            if not node:
                return 0

            left_hight = get_hight(node.left)
            right_hight = get_hight(node.right)
            
            # update the max diameter
            max_road = max(max_road, left_hight + right_hight)
            
            return 1 + max(left_hight, right_hight)
        
        get_hight(root)
        return max_road

        